src/main/java/io/github/genomicdatainfrastructure/discovery/datasets/application/usecases/SearchDatasetsQuery.java

```diff
@@ -6,64 +6,78 @@
-import io.github.genomicdatainfrastructure.discovery.datasets.application.ports.RecordsCountCollector;
 import io.github.genomicdatainfrastructure.discovery.model.DatasetSearchQuery;
 import io.github.genomicdatainfrastructure.discovery.model.DatasetsSearchResponse;
 import jakarta.enterprise.context.ApplicationScoped;
 import jakarta.enterprise.inject.Instance;
 import jakarta.inject.Inject;
 import lombok.RequiredArgsConstructor;
 
-import java.util.List;
+import java.util.HashMap;
+import java.util.Map;
 import java.util.Objects;
 
-import static java.util.Optional.ofNullable;
+import static java.lang.Math.min;
+import static java.util.Objects.nonNull;
 
 @ApplicationScoped
 @RequiredArgsConstructor(onConstructor = @__(@Inject))
 public class SearchDatasetsQuery {
 
     private final DatasetsRepository repository;
     private final Instance<DatasetIdsCollector> collectors;
-    private final RecordsCountCollector recordsCountCollector;
 
     public DatasetsSearchResponse execute(DatasetSearchQuery query, String accessToken) {
-        var datasetIds = collectors
+        var datasetIdsByRecordCount = collectors
                 .stream()
                 .map(collector -> collector.collect(query, accessToken))
                 .filter(Objects::nonNull)
                 .reduce(this::findIdsIntersection)
-                .orElse(List.of());
+                .orElseGet(Map::of);
 
-        var datasets = repository.search(datasetIds,
+        var datasets = repository.search(datasetIdsByRecordCount.keySet(),
                 query.getSort(),
                 query.getRows(),
                 query.getStart(),
                 accessToken);
 
-        var potentialRecordsCounts = ofNullable(recordsCountCollector.collectRecordsCount(query,
-                accessToken));
-
         var enhancedDatasets = datasets
                 .stream()
                 .map(dataset -> dataset
                         .toBuilder()
-                        .recordsCount(potentialRecordsCounts
-                                .map(recordsCounts -> recordsCounts.get(dataset.getIdentifier()))
-                                .orElse(null))
+                        .recordsCount(datasetIdsByRecordCount.get(dataset.getIdentifier()))
                         .build())
                 .toList();
 
         return DatasetsSearchResponse
                 .builder()
-                .count(datasetIds.size())
+                .count(datasetIdsByRecordCount.size())
                 .results(enhancedDatasets)
                 .build();
     }
 
-    private List<String> findIdsIntersection(List<String> a, List<String> b) {
-        return a.stream()
-                .filter(b::contains)
-                .toList();
+    private Map<String, Integer> findIdsIntersection(
+            Map<String, Integer> a,
+            Map<String, Integer> b
+    ) {
+        var newMap = new HashMap<String, Integer>();
+        for (var entryA : a.entrySet()) {
```

Comment:
I am not sure how many datasets will be present at the end, but performance goes down with increasing datasets. why not using `parallelStream`? 
Moreover, sonar did not complain about cognitive comlexity?

Reply:
Hey @zalborzi thanks for this question, indeed O(N) could be improved, but it is still not bad, counting that it stays O(N).

I agree it could be a problem with a couple of hundred thousands of object, but by experience `parallelStream` may introduce worst issues than slowness. I would introduce other solutions to improve performance, but for now let the performance issue happen before considering fix it, because we don't know where it will happen.